cw1/cw1.py

```diff
@@ -0,0 +1,184 @@
+import numpy as np
+import matplotlib.pyplot as plt
+
+# Define functions
+
+def f(x):
+    return 10*x**4 + 3*x**3 - 30*x**2 + 10*x
+
+def g(x1, x2):
+    return (x1 - 2)**4 + (x2 + 3)**4 + 2 * (x1 - 2)**2 * (x2 + 3)**2
+
+# Define gradients
+
+def gradient_f(x):
+    return 40*x**3 + 9*x**2 - 60*x + 10
+
+def gradient_g(x1, x2):
+    df_dx1 = 4*(x1 - 2)**3 + 4*(x1 - 2)*(x2 + 3)**2
+    df_dx2 = 4*(x2 + 3)**3 + 4*(x2 + 3)*(x1 - 2)**2
+    return np.array([df_dx1, df_dx2])
+
+# Gradient Descent Implementation
+def gradient_descent(fun, gradient, initial_point, beta, max_iterations=10000, treshold=1e-7):
+    x = np.array(initial_point, dtype=float)  # Ensure x is a NumPy array
+    path = [x.copy()]
+    
+    for iteration in range(max_iterations):
+        if x.ndim == 0:  # If x is a scalar
```

Comment:
podpowiedź: `x` jest typu `np.ndarray`, więc można to też zrobić w następujący sposób (zadziała dla skalarów i wektorów): `gradient(*x.reshape(-1))`

`x.reshape(-1)` przekształca dowolną NumPy array w 1-wymiarowy wektor (`-1` jako kształt oznacza, że ma być jeden wymiar o automatycznie dopasowanym rozmiarze), więc skalar `x` przekształci w wektor `[x]`.